hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java

```diff
@@ -0,0 +1,694 @@
+    // This should have been checked in BasicRootedOzoneFileSystem#rename
+    // already via regular call path unless bypassed.
+    if (!ofsPath.isInSameBucketAs(ofsNewPath)) {
+      throw new IOException("Cannot rename a key to a different bucket");
+    }
+
+    OzoneBucket bucket = getBucket(ofsPath, false);
+    String key = ofsPath.getKeyName();
+    String newKey = ofsNewPath.getKeyName();
+    bucket.renameKey(key, newKey);
+  }
+
+  /**
+   * Package-private helper function to reduce calls to getBucket().
+   * @param bucket Bucket to operate in.
+   * @param path Existing key path.
+   * @param newPath New key path.
+   * @throws IOException IOException from bucket.renameKey().
+   */
+  void renamePath(OzoneBucket bucket, String path, String newPath)
+      throws IOException {
+    incrementCounter(Statistic.OBJECTS_RENAMED);
+    OFSPath ofsPath = new OFSPath(path);
+    OFSPath ofsNewPath = new OFSPath(newPath);
+    // No same-bucket policy check here since this call path is controlled
+    String key = ofsPath.getKeyName();
+    String newKey = ofsNewPath.getKeyName();
+    bucket.renameKey(key, newKey);
+  }
+
+  /**
+   * Helper method to create an directory specified by key name in bucket.
+   *
+   * @param pathStr path to be created as directory
+   * @return true if the key is created, false otherwise
+   */
+  @Override
+  public boolean createDirectory(String pathStr) throws IOException {
+    LOG.trace("creating dir for path: {}", pathStr);
+    incrementCounter(Statistic.OBJECTS_CREATED);
+    OFSPath ofsPath = new OFSPath(pathStr);
+    // Volume name unspecified, return failure
+    if (ofsPath.getVolumeName().length() == 0) {
+      return false;
+    }
+    // Handle where only volume is specified in pathStr
+    if (ofsPath.getBucketName().length() == 0) {
+      objectStore.createVolume(ofsPath.getVolumeName());
+      return true;
+    }
+    String keyStr = ofsPath.getKeyName();
+    try {
+      OzoneBucket bucket = getBucket(ofsPath, true);
+      // if keyStr is empty, it indicates that only volume or volume+bucket is
+      // given in pathStr, so getBucket() above should've handled the creation
+      // of volume/bucket already.
+      if (keyStr != null && keyStr.length() > 0) {
+        bucket.createDirectory(keyStr);
+      }
+    } catch (OMException e) {
+      if (e.getResult() == OMException.ResultCodes.FILE_ALREADY_EXISTS) {
+        throw new FileAlreadyExistsException(e.getMessage());
+      }
+      throw e;
+    }
+    return true;
+  }
+
+  /**
+   * Helper method to delete an object specified by key name in bucket.
+   *
+   * @param path path to a key to be deleted
+   * @return true if the key is deleted, false otherwise
+   */
+  @Override
+  public boolean deleteObject(String path) {
+    LOG.trace("issuing delete for path to key: {}", path);
+    OFSPath ofsPath = new OFSPath(path);
+    String keyName = ofsPath.getKeyName();
```

Comment:
check keyName is empty?

Reply:
Do we support delete volume/bucket via OFS?

Reply:
keyName check done in ac0cd77f9c22c19c7442f7e248ab401bbb9ffb73.

Reply:
On supporting delete volume/bucket though, this is a good question. I'd argue it is weird to support delete volume/bucket via `rm` interface but from a flat filesystem point of view it should be a feature. What's you opinion? @arp7

Reply:
Let's follow up that on separate JIRA.

Reply:
Agree